File: banking/src/main/java/com/sina/banking/models/Account.java
package com.sina.banking.models;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;

@Entity
@Table(name = "accounts")
@EntityListeners(AuditingEntityListener.class)
@Setter @Getter
public class Account {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(name = "account_number", unique = true, nullable = false)
    private Long accountNumber;

    @Enumerated(EnumType.STRING)
    @Column(name = "account_type", nullable = false)
    private AccountType type;

    @Column(name = "currency", nullable = false)
    private String currency;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private AccountStatus status;

    // Optimistic lock for direct mutations to this row (freeze/close/activate). Not what protects
    // withdraw/transfer's balance check, though - those never write to this row directly, so they
    // rely on a pessimistic row lock instead (see AccountRepository.findByIdForUpdate).
    @Column(name = "version")
    @Version
    private Integer version;

    @CreatedDate
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;

    public Account() {}

    public Account(User user, Long accountNumber, AccountType type, String currency) {
        this.user = user;
        this.accountNumber = accountNumber;
        this.type = type;
        this.currency = currency;
        this.status = AccountStatus.ACTIVE;
    }

    public void freeze() {
        this.status = AccountStatus.FROZEN;
    }

    public void close() {
        this.status = AccountStatus.CLOSED;
    }

    public void active() {
        this.status = AccountStatus.ACTIVE;
    }
}

File: reporting-service/app/models/account.py
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy import ForeignKey
from enum import StrEnum
from app.db.database import Base


import datetime

class AccountType(StrEnum):
    CHECKING = 'CHECKING'
    SAVINGS = 'SAVINGS'
    SYSTEM = 'SYSTEM'

class AccountStatus(StrEnum):
    ACTIVE = 'ACTIVE'
    FROZEN = 'FROZEN'
    CLOSED = 'CLOSED'

class Account(Base):
    __tablename__ = 'accounts'

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(ForeignKey('users.id'), nullable=False)
    user: Mapped['User'] = relationship(back_populates="accounts")
    account_number: Mapped[int] = mapped_column(name='account_number', nullable=False)
    account_type: Mapped[AccountType] = mapped_column(name='account_type', nullable=False)
    currency: Mapped[str] = mapped_column(name='currency', nullable=False)
    account_status: Mapped[AccountStatus] = mapped_column(name='status', nullable=False)
    version: Mapped[int] = mapped_column(name='version')
    created_at: Mapped[datetime.datetime] = mapped_column(name='created_at')
    updated_at: Mapped[datetime.datetime] = mapped_column(name='updated_at')
    ledger_entries: Mapped[list['LedgerEntry']] = relationship(back_populates='account')

